Completed homework 22 sprint)

RURURuina opened the pull request:
Очень много сил было вложено)))
Спасибо за проверку!

RURURuina committed:
Completed homework 22 sprint)
Очень много сил было вложено)))
Спасибо за проверку!

sukhoikms27 reviewed (commented):
Креш на старте, с чистой установкой:
```
FATAL EXCEPTION: main
Process: com.practicum.playlistmaker, PID: 15445
java.lang.IllegalStateException: A migration from 3 to 2 was required but not found. Please provide the necessary Migration path via RoomDatabase.Builder.addMigration(Migration ...) or allow for destructive migrations via one of the RoomDatabase.Builder.fallbackToDestructiveMigration* methods.
	at androidx.room.RoomOpenHelper.onUpgrade(RoomOpenHelper.kt:109)
	at androidx.room.RoomOpenHelper.onDowngrade(RoomOpenHelper.kt:122)
	at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.onDowngrade(FrameworkSQLiteOpenHelper.kt:277)
	at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:414)
	at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:316)
	at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.getWritableOrReadableDatabase(FrameworkSQLiteOpenHelper.kt:232)
	at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.innerGetDatabase(FrameworkSQLiteOpenHelper.kt:190)
	at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.getSupportDatabase(FrameworkSQLiteOpenHelper.kt:151)
	at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper.getWritableDatabase(FrameworkSQLiteOpenHelper.kt:104)
	at androidx.room.RoomDatabase.inTransaction(RoomDatabase.kt:632)
	at androidx.room.RoomDatabase.assertNotSuspendingTransaction(RoomDatabase.kt:451)
	at androidx.room.RoomDatabase.query(RoomDatabase.kt:480)
	at androidx.room.util.DBUtil.query(DBUtil.kt:75)
	at com.practicum.playlistmaker.mediateka.data.db.dao.FavoriteTrackDao_Impl$5.call(FavoriteTrackDao_Impl.java:156)
	at com.practicum.playlistmaker.mediateka.data.db.dao.FavoriteTrackDao_Impl$5.call(FavoriteTrackDao_Impl.java:152)
	at androidx.room.CoroutinesRoom$Companion$createFlow$1$1$1.invokeSuspend(CoroutinesRoom.kt:129)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
	at java.lang.Thread.run(Thread.java:1012)
	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@5e08bf1, Dispatchers.Main.immediate]
Sending signal. PID: 15445 SIG: 9
```

RURURuina committed:
Completed homework 22 sprint)
Очень много сил было вложено)))
Добавил пересборку, в 23 спринте допилю миграцию
Спасибо за проверку!

sukhoikms27 commented on app/src/main/java/com/practicum/playlistmaker/mediateka/ui/FragmentPlaylists.kt:
```diff
@@ -23,6 +30,57 @@ class FragmentPlaylists : Fragment() {
+
+            playlistsRecyclerView.adapter?.notifyDataSetChanged()
+
+            binding.playlistsRecyclerView.visibility = View.VISIBLE
```
Рекомендация:
для удобства можно пользоваться экстеншеном: `.isVisible = true/false`

sukhoikms27 commented on app/src/main/java/com/practicum/playlistmaker/mediateka/ui/adapters/PlaylistViewHolder.kt:
```diff
@@ -0,0 +1,21 @@
+        fun bind(playlist: Playlist) {
+            with(binding) {
+                playlistName.text = playlist.playlistName
+                tracksCount.text = playlist.tracksCount.toString() + " " + TrackWordConvertor.getTrackCountString(playlist.tracksCount, itemView)
```
Рекомендация:
Конкатенацию строк через оператор `+` лучше не использовать в большом количестве, так как она каждый раз создает объект `String`, особенно плохо так делать в цикле. Лучше многократную конкатенация вынести в один объект `StringBuilder`

sukhoikms27 commented on app/src/main/java/com/practicum/playlistmaker/player/ui/AudioPlayerFragment.kt:
```diff
@@ -0,0 +1,244 @@
+    private val binding get() = _binding!!
+
+    private val args: AudioPlayerFragmentArgs by navArgs()
+    //private lateinit var playButton: ImageView
```
Реккомендация:
Закоментированный код лучше сразу удалять. Хранить закоментированный код в проекте - плохая практика, в случае чего этот код можно восстановить с помощью истории гита

sukhoikms27 commented on app/src/main/java/com/practicum/playlistmaker/player/ui/AudioPlayerFragment.kt:
```diff
@@ -0,0 +1,244 @@
+    //private lateinit var favoriteButton: ImageView
+    private var adapter: BottomSheetPlaylistAdapter? = null
+    private lateinit var bottomSheet: BottomSheetBehavior<LinearLayout>
+    companion object {
```
Рекомендация:
`companion object` принято объявлять внизу класса согласно Kotlin Coding Convention (https://kotlinlang.org/docs/coding-conventions.html#class-layout)

sukhoikms27 commented on app/src/main/java/com/practicum/playlistmaker/utils/TrackWordConvertor.kt:
```diff
@@ -0,0 +1,20 @@
+            val remainder100 = count % 100
+
+            return when {
+                remainder100 in 11..19 -> view.context.getString(R.string.track_count_5)
```
хорошее решение проблемы, но лучше посмотреть в сторону plurals:
https://developer.android.com/guide/topics/resources/string-resource#Plurals

RURURuina commented:
Спасибо за проверку, внесу рекомендационные правки
